File: week_25/02_mutex.rb
#!/usr/bin/env ruby

# 02. Mutex in Ruby - Synchronization

puts "=== Ruby Mutex - Synchronization ==="
puts

# Problem: Race condition
puts "1. Race condition (WITHOUT mutex):"
counter = 0
threads = 10.times.map do
  Thread.new do
    1000.times { counter += 1 }
  end
end
threads.each(&:join)
puts "  Expected: 10000"
puts "  Got:      #{counter}"  # Може бути менше через race condition!
puts

# Solution: Mutex
puts "2. Thread-safe (WITH mutex):"
mutex = Mutex.new
counter = 0

threads = 10.times.map do
  Thread.new do
    1000.times do
      mutex.synchronize do
        counter += 1
      end
    end
  end
end

threads.each(&:join)
puts "  Expected: 10000"
puts "  Got:      #{counter}"  # Завжди 10000!
puts

# Mutex methods
puts "3. Mutex methods:"
mutex = Mutex.new

puts "  Locked? #{mutex.locked?}"

mutex.lock
puts "  Locked? #{mutex.locked?}"

mutex.unlock
puts "  After unlock? #{mutex.locked?}"
puts

# Try lock (non-blocking)
puts "4. Try lock (non-blocking):"
mutex = Mutex.new

if mutex.try_lock
  puts "  Lock acquired!"
  mutex.unlock
else
  puts "  Lock failed"
end
puts

# Real-world example: Bank account
puts "5. Real-world: Bank account"
class BankAccount
  def initialize(balance = 0)
    @balance = balance
    @mutex = Mutex.new
  end

  def balance
    @mutex.synchronize { @balance }
  end

  def deposit(amount)
    @mutex.synchronize do
      @balance += amount
    end
  end

  def withdraw(amount)
    @mutex.synchronize do
      return false if @balance < amount
      @balance -= amount
      true
    end
  end
end

account = BankAccount.new(1000)

# Multiple threads accessing account
threads = []
10.times do
  threads << Thread.new { account.deposit(100) }
  threads << Thread.new { account.withdraw(50) }
end

threads.each(&:join)
puts "  Final balance: $#{account.balance}"
puts

puts "✅ Ruby Mutex complete"

# Key points:
# - Mutex.new to create
# - mutex.synchronize { } for safe access
# - mutex.lock / mutex.unlock (manual)
# - mutex.try_lock (non-blocking)
# - Prevents race conditions

File: week_25/02_mutex.go
package main

import (
	"fmt"
	"sync"
)

// 02. Mutex in Go - Synchronization

func main() {
	fmt.Println("=== Go Mutex - Synchronization ===")
	fmt.Println()

	// Problem: Race condition
	fmt.Println("1. Race condition (WITHOUT mutex):")
	counter := 0
	var wg sync.WaitGroup

	for i := 0; i < 10; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := 0; j < 1000; j++ {
				counter++ // Race condition!
			}
		}()
	}

	wg.Wait()
	fmt.Printf("  Expected: 10000\n")
	fmt.Printf("  Got:      %d\n", counter) // Може бути менше!
	fmt.Println()

	// Solution: Mutex
	fmt.Println("2. Thread-safe (WITH mutex):")
	counter = 0
	var mu sync.Mutex

	for i := 0; i < 10; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := 0; j < 1000; j++ {
				mu.Lock()
				counter++
				mu.Unlock()
			}
		}()
	}

	wg.Wait()
	fmt.Printf("  Expected: 10000\n")
	fmt.Printf("  Got:      %d\n", counter) // Завжди 10000!
	fmt.Println()

	// RWMutex (Read-Write Mutex)
	fmt.Println("3. RWMutex (multiple readers, single writer):")
	var rwmu sync.RWMutex
	data := 0

	// Multiple readers
	for i := 0; i < 5; i++ {
		wg.Add(1)
		go func(n int) {
			defer wg.Done()
			rwmu.RLock()
			fmt.Printf("  Reader %d: %d\n", n, data)
			rwmu.RUnlock()
		}(i)
	}

	// Single writer
	wg.Add(1)
	go func() {
		defer wg.Done()
		rwmu.Lock()
		data = 42
		fmt.Println("  Writer: updated to 42")
		rwmu.Unlock()
	}()

	wg.Wait()
	fmt.Println()

	// Real-world example: Bank account
	fmt.Println("4. Real-world: Bank account")
	account := NewBankAccount(1000)

	// Multiple goroutines accessing account
	for i := 0; i < 10; i++ {
		wg.Add(2)
		go func() {
			defer wg.Done()
			account.Deposit(100)
		}()
		go func() {
			defer wg.Done()
			account.Withdraw(50)
		}()
	}

	wg.Wait()
	fmt.Printf("  Final balance: $%d\n", account.Balance())
	fmt.Println()

	fmt.Println("✅ Go Mutex complete")
}

// BankAccount with mutex
type BankAccount struct {
	balance int
	mu      sync.Mutex
}

func NewBankAccount(balance int) *BankAccount {
	return &BankAccount{balance: balance}
}

func (a *BankAccount) Balance() int {
	a.mu.Lock()
	defer a.mu.Unlock()
	return a.balance
}

func (a *BankAccount) Deposit(amount int) {
	a.mu.Lock()
	defer a.mu.Unlock()
	a.balance += amount
}

func (a *BankAccount) Withdraw(amount int) bool {
	a.mu.Lock()
	defer a.mu.Unlock()
	if a.balance < amount {
		return false
	}
	a.balance -= amount
	return true
}

// Key points:
// - sync.Mutex for exclusive access
// - mu.Lock() / mu.Unlock()
// - defer mu.Unlock() pattern
// - sync.RWMutex for read-heavy workloads
// - No GIL - true parallelism
